### PR TITLE
Pass all specs to ingestion by file

### PR DESCRIPTION
**Pull request by tims**
Ingestion now takes a `--workspace` option which should be a path to a directory containing importJobSpecs.yaml file.
This yaml file should align with a new `feast.types.ImportJobSpecs` protobuf which bundles all the used specs by an import job into a single proto. 
The workspace directory is also used to write out the errors if no errors storageSpec is provided.

This importJobSpecs.yaml is be generated by core before starting the job. It is not intended to be written by humans (thought it's useful for testing). It is intended to make it easier for the core api to make all the necessary (and only the necessary) specs available to ingestion. Without Ingestion needing to call back to core. This make deploying feast simpler. It also could make debugging what happened with a job a lot easier.

An alternative approach was to have spec in a separate file, but I think a single file this is easier and neater.

The proto looks like:
```
message ImportJobSpecs {
  string jobId = 1;
  feast.specs.ImportSpec importSpec = 2;
  repeated feast.specs.EntitySpec entitySpecs = 3;
  repeated feast.specs.FeatureSpec featureSpecs = 4;
  repeated StorageSpec servingStorageSpecs = 5;
  repeated StorageSpec warehouseStorageSpecs = 6;
  StorageSpec errorsStorageSpec = 7;
}
```

Here is an example of the corresponding yaml:
```
importSpec: 
  type: file.csv
  options:
    path: /tmp/file.csv
  entities:
    - testEntity
  schema:
    entityIdColumn: id
    timestampValue: 2018-09-25T00:00:00.000Z
    fields:
      - name: timestamp
      - name: id
      - featureId: testEntity.none.testInt64
servingStorageSpecs:
  - id: TEST_SERVING
    type: serving.mock
    options: {}
warehouseStorageSpecs:
  - id: TEST_WAREHOUSE
    type: warehouse.mock
    options: {}
entitySpecs:
  - name: testEntity
    description: This is a test entity
    tags: []
featureSpecs:
  - id: testEntity.day.testInt64
    entity: testEntity
    granularity: DAY
    name: testInt64
    owner: feast@example.com
    description: This is test feature of type integer
    uri: https://example.com/
    valueType: INT64
    tags: []
    options: {}
```

### CODE REVIEW

**Comment by tims** [comment]
Also worth noting, ImportJobSpec is not a proto for public consumption, just for communicating between core and ingestion. 

It is a great idea for us to find a way for users to define multiple feature specs and an import spec all in one yaml file, but that is a completely independent problem (and would suit a much more concise format than this one is). So let's stay on topic and not get caught on that.

**Comment by tims** [comment]
@pradithya ready for review

**Comment by tims** [comment]
I original include additional changes that limited to single data store for serving and warehouse, but I removed that to make this easier to review. 

Instead: the additional core changes include:

Core can now also be configured with the following additional properties:
```
feast.store.serving.type = ${STORE_SERVING_TYPE:}
feast.store.serving.options = ${STORE_SERVING_OPTIONS:{}}
feast.store.warehouse.type = ${STORE_WAREHOUSE_TYPE:}
feast.store.warehouse.options = ${STORE_WAREHOUSE_OPTIONS:{}}
feast.store.errors.type = ${STORE_ERRORS_TYPE:}
feast.store.errors.options = ${STORE_ERRORS_OPTIONS:{}}
```

If present, these will be substituted when a feature spec is applied that does not provide any store id.
The serving and warehouse stores are registered on startup, the same way a use might register a store. So if you only want these stores, you never need to provide any store references in the feature specs.

The following core properties are now mandatory:
```
feast.jobs.workspace=${JOB_WORKSPACE}
```
And `feast.jobs.coreUri` has been removed as it is no longer necessary for ingestion to know the uri of core.

This workspace must point to a directory in core and ingestion have read and write access. Core will create a subdirectory per job and provide that as the ingestion job's workspace directory, then write a importJobSpecs.yaml file to it. Ingestion will read that file to configure start the beam pipeline.

Additionally, if the importErrorsSpecs proto does not include a errorsStorageSpec in it, ingestion will default to writing errors with type "file.json" to the it's workspace/errors directory.

Question for reviewers. Core and ingestion overload the name workspace, for core it means the base path, for ingestion it means the {base path}/{jobId}. Is this acceptable or should we change configuration of core to be baseWorkspace?

**Comment by tims** [comment]
/hold

updating the helm chart

**Comment by tims** [comment]
/hold

there's an issue where core is not using google-cloud-nio to access the gcs filesystem.

**Comment by tims** [comment]
I'm getting these errors:
```
   extendedStackTrace: [17]    
   localizedMessage:  "Provider "gs" not installed"    
   message:  "Provider "gs" not installed"    
   name:  "java.nio.file.FileSystemNotFoundException"    
```
```
   extendedStackTrace: [
    0: {
     class:  "java.nio.file.Paths"      
     exact:  false      
     file:  "Paths.java"      
     line:  147      
     location:  "?"      
     method:  "get"      
     version:  "1.8.0_181"      
    }
    1: {
     class:  "feast.core.util.PathUtil"      
     exact:  false      
     file:  "PathUtil.java"      
     line:  33      
     location:  "classes!/"      
     method:  "getPath"      
     version:  "?"      
    }
```

This does not make sense to ,me since I've added the google-cloud-nio library and it passes the tests. I can reproduce the error in the tests if I remove it from the pom.xml. :( 

Any suggestions for why this might not be included in build?

**Comment by tims** [comment]
out of paranoia I confirmed that it does exist in BOOT-INF/lib/google-cloud-nio-0.83.0-alpha.jar if I unzip the core jar. hmmmm... 🤔

**Comment by pradithya** [comment]
```
feast.store.serving.type = ${STORE_SERVING_TYPE:}
feast.store.serving.options = ${STORE_SERVING_OPTIONS:{}}
feast.store.warehouse.type = ${STORE_WAREHOUSE_TYPE:}
feast.store.warehouse.options = ${STORE_WAREHOUSE_OPTIONS:{}}
feast.store.errors.type = ${STORE_ERRORS_TYPE:}
feast.store.errors.options = ${STORE_ERRORS_OPTIONS:{}}
```

Are these mandatory? I am not able to start core without setting them

**Comment by pradithya** [comment]
I got following error when trying to submit job and the workspace is in GCS

```
Optional[Exception in thread "main" java.lang.UnsupportedOperationException: GCS objects aren't available locally
	at com.google.cloud.storage.contrib.nio.CloudStoragePath.toFile(CloudStoragePath.java:303)
	at feast.ingestion.util.ProtoUtil.decodeProtoYamlFile(ProtoUtil.java:37)
	at feast.ingestion.config.ImportJobSpecsSupplier.create(ImportJobSpecsSupplier.java:45)
	at feast.ingestion.config.ImportJobSpecsSupplier.get(ImportJobSpecsSupplier.java:57)
	at feast.ingestion.ImportJob.mainWithResult(ImportJob.java:121)
	at feast.ingestion.ImportJob.main(ImportJob.java:109)]
	at feast.core.job.direct.DirectRunnerJobManager.runProcess(DirectRunnerJobManager.java:113)
```

**Comment by tims** [comment]
> I got following error when trying to submit job and the workspace is in GCS
> 
> ```
> Optional[Exception in thread "main" java.lang.UnsupportedOperationException: GCS objects aren't available locally
> 	at com.google.cloud.storage.contrib.nio.CloudStoragePath.toFile(CloudStoragePath.java:303)
> 	at feast.ingestion.util.ProtoUtil.decodeProtoYamlFile(ProtoUtil.java:37)
> 	at feast.ingestion.config.ImportJobSpecsSupplier.create(ImportJobSpecsSupplier.java:45)
> 	at feast.ingestion.config.ImportJobSpecsSupplier.get(ImportJobSpecsSupplier.java:57)
> 	at feast.ingestion.ImportJob.mainWithResult(ImportJob.java:121)
> 	at feast.ingestion.ImportJob.main(ImportJob.java:109)]
> 	at feast.core.job.direct.DirectRunnerJobManager.runProcess(DirectRunnerJobManager.java:113)
> ```

This is strange, because I got a similar error from core, when it was trying to write the specs, you got it ingestion when it was trying to read them.

**Comment by tims** [comment]
> ```
> feast.store.serving.type = ${STORE_SERVING_TYPE:}
> feast.store.serving.options = ${STORE_SERVING_OPTIONS:{}}
> feast.store.warehouse.type = ${STORE_WAREHOUSE_TYPE:}
> feast.store.warehouse.options = ${STORE_WAREHOUSE_OPTIONS:{}}
> feast.store.errors.type = ${STORE_ERRORS_TYPE:}
> feast.store.errors.options = ${STORE_ERRORS_OPTIONS:{}}
> ```
> Are these mandatory? I am not able to start core without setting them

They are both optional. Just fixed.
A serving store still needs to be registered before feature specs can pass validation

**Comment by pradithya** [comment]
> This is strange, because I got a similar error from core, when it was trying to write the specs, you got it ingestion when it was trying to read them.

For me, ImportJobSpec is successfully uploaded to GCS, but ingestion is unable to access it with above error.

Note that, I have to change following line when passing the workspace

```
    commands.add(option("workspace", workspace.toUri().toString()));
```

File: core/src/test/java/feast/core/job/dataflow/DataflowJobManagerTest.java [hunk 1]
```diff
@@ -25,57 +25,67 @@
...
 
   @Test
   public void shouldBuildProcessBuilderWithCorrectOptions() {
     ImportSpec importSpec = ImportSpec.newBuilder().setType("file").build();
-    ProcessBuilder pb = dfJobManager.getProcessBuilder(importSpec, "test");
+    String jobName = "test";
+    ImportJobSpecs importJobSpecs = ImportJobSpecs.newBuilder().setJobId(jobName)
+        .setImportSpec(importSpec).build();
+
+    ProcessBuilder pb = dfJobManager.getProcessBuilder(importJobSpecs, Paths.get("/tmp/foobar"));
```
Comment (pradithya) [inline core/src/test/java/feast/core/job/dataflow/DataflowJobManagerTest.java]: Can you add test where the workspace is in GCS?

File: ingestion/src/main/java/feast/ingestion/config/ImportJobSpecsSupplier.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    if (!Strings.isNullOrEmpty(workspace)) {
+      this.importJobSpecsPath = PathUtil.getPath(workspace).resolve(IMPORT_JOB_SPECS_FILENAME);
+    }
+  }
+
+  private ImportJobSpecs create() {
+    try {
+      if (importJobSpecsPath != null) {
+        return ProtoUtil
+            .decodeProtoYamlFile(importJobSpecsPath, ImportJobSpecs.getDefaultInstance());
```
Comment (pradithya) [inline ingestion/src/main/java/feast/ingestion/config/ImportJobSpecsSupplier.java]: `decodeProtoYamlFile` should be able to handle the case when  `importJobSpecsPath` is in GCS.

File: core/src/main/java/feast/core/job/direct/DirectRunnerJobManager.java [hunk 3]
```diff
@@ -66,25 +67,19 @@ public void abortJob(String extId) {
...
+  public ProcessBuilder getProcessBuilder(ImportJobSpecs importJobSpecs, Path workspace) {
     Map<String, String> options =
         TypeConversion.convertJsonStringToMap(defaults.getImportJobOptions());
     List<String> commands = new ArrayList<>();
     commands.add("java");
     commands.add("-jar");
     commands.add(defaults.getExecutable());
-    commands.add(option("jobName", jobId));
+    commands.add(option("jobName", importJobSpecs.getJobId()));
+    commands.add(option("workspace", workspace.toString()));
```
Comment (pradithya) [inline core/src/main/java/feast/core/job/direct/DirectRunnerJobManager.java]: It has to be `workspace.toUri(),toString()` to handle remote file

**Pull request merged**